src/libmongoc/src/mongoc/mongoc-async-cmd.c

```diff
@@ -61,22 +64,32 @@ static const _mongoc_async_cmd_phase_t gMongocCMDPhases[] = {
 
 #ifdef MONGOC_ENABLE_SSL
 int
-mongoc_async_cmd_tls_setup (mongoc_stream_t *stream, int *events, void *ctx, int32_t timeout_msec, bson_error_t *error)
+mongoc_async_cmd_tls_setup (mongoc_stream_t *stream, int *events, void *ctx, mlib_timer deadline, bson_error_t *error)
 {
    mongoc_stream_t *tls_stream;
    const char *host = (const char *) ctx;
    int retry_events = 0;
 
-
    for (tls_stream = stream; tls_stream->type != MONGOC_STREAM_TLS;
         tls_stream = mongoc_stream_get_base_stream (tls_stream)) {
    }
 
+   const bool use_non_blocking =
 #if defined(MONGOC_ENABLE_SSL_OPENSSL) || defined(MONGOC_ENABLE_SSL_SECURE_CHANNEL)
-   /* pass 0 for the timeout to begin / continue non-blocking handshake */
-   timeout_msec = 0;
+      true
+#else
+      false
 #endif
-   if (mongoc_stream_tls_handshake (tls_stream, host, timeout_msec, &retry_events, error)) {
+      ;
+
+   // Try to do a non-blocking operation, if our backend allows it
+   const mlib_duration_rep_t remain_ms = //
+      use_non_blocking
+         // Pass 0 for the timeout to begin / continue a non-blocking handshake
+         ? 0
+         // Otherwise, use the deadline
+         : mlib_milliseconds_count (mlib_timer_remaining (deadline));
+   if (mongoc_stream_tls_handshake (tls_stream, host, remain_ms, &retry_events, error)) {
```

Comment:
```
warning: conversion from ‘mlib_duration_rep_t’ {aka ‘long int’} to ‘int32_t’ {aka ‘int’} may change value [-Wconversion]
   92 |    if (mongoc_stream_tls_handshake (tls_stream, host, remain_ms, &retry_events, error)) {
      |                                                       ^~~~~~~~~
```

Looks like [this cast](https://github.com/mongodb/mongo-c-driver/blob/694fbf3da960478f4ec5589df625549a72856777/src/libmongoc/src/mongoc/mongoc-async-cmd.c#L276) ended up moving here, where the public API `mongoc_stream_tls_handshake` unfortunately requires an `int32_t`). Suggest restoring the check (clamp?) + explicit cast to `int32_t` here.